From c16ba077cdab1a7a4424063e165a4d373595c937 Mon Sep 17 00:00:00 2001
From: Karthik M <quic_karm@quicinc.com>
Date: Fri, 13 Dec 2024 15:48:04 +0530
Subject: [PATCH] UPSTREAM:PROTOCOL:wifi: Add ieee_link_id param during vdev
 create

Link-id in 2GHz/5GHz beacon perSta profile is 0 when CSA triggered at 6G Txvap.
It should be 2 as 6GHz is mapped to link-id 2.

During channel switch, per sta profile information will be populated based
on the hw_link_id. In upstream design, ieee_link_id will not be same as
hw_link_id. ieee_link_id for a link will not be same for
all vdev in a given pdev.

For example, consider Three MLD's
MLD 1 -  5GHz and 6GHz  (2link MLO)
MLD 2 -  6GHz           (1link SLO)
MLD 3 -  5GHz           (1link SLO)

Here the link ID assignment will be,

MLD 1 -  5GHz link (link id 0) and 6GHz link (link id 1)  -> 2 link MLO
MLD 2 -  6GHz link SLO (link id 0)                        -> 1 link SLO
MLD 3 -  5GHz link SLO (link id 0)                        -> 1 link SLO

FW need ieee_link_id through WMI during vdev start from
Host to correctly populate the values in per sta profile IE.
Add support to advertise ieee_link_id of both partner link and
self in wmi_vdev_start_mlo_params TLV

To assure backward compatability, Two flags has been introduced
for self and partner link. FW will parse the parameters based on these
flags.
1. ATH12K_WMI_FLAG_MLO_IEEE_LINK_IDX_VALID BIT(18)
2. ATH12K_WMI_FLAG_MLO_IEEE_LINK_IDX_VALID_PARTNER BIT(19)

Patch-work: None

Patch-dependency: FW

Co-developed-by: Hari Naraayana Desikan Kannan <quic_hnaraaya@quicinc.com>
Signed-off-by: Hari Naraayana Desikan Kannan <quic_hnaraaya@quicinc.com>
Signed-off-by: Karthik M <quic_karm@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/mac.c |  1 +
 drivers/net/wireless/ath/ath12k/wmi.c | 19 ++++++++++++++-----
 drivers/net/wireless/ath/ath12k/wmi.h |  2 ++
 3 files changed, 17 insertions(+), 5 deletions(-)

--- a/drivers/net/wireless/ath/ath12k/mac.c
+++ b/drivers/net/wireless/ath/ath12k/mac.c
@@ -14070,6 +14070,8 @@ ath12k_mac_mlo_get_vdev_args(struct ath1
 	ml_arg->link_add = true;
 	ml_arg->mlo_bridge_link = ath12k_mac_is_bridge_vdev(arvif);
 
+	ml_arg->ieee_link_id = arvif->link_id;
+
 	/* TODO check assoc and mcast vdev for AP mode */
 
 	partner_info = ml_arg->partner_info;
@@ -14092,6 +14094,7 @@ ath12k_mac_mlo_get_vdev_args(struct ath1
 
 		partner_info->vdev_id = arvif_p->vdev_id;
 		partner_info->hw_link_id = arvif_p->ar->pdev->hw_link_id;
+		partner_info->logical_link_idx = arvif_p->link_id;
 		partner_info->mlo_bridge_link = ath12k_mac_is_bridge_vdev(arvif_p);
 		ml_arg->num_partner_links++;
 		partner_info++;
--- a/drivers/net/wireless/ath/ath12k/wmi.c
+++ b/drivers/net/wireless/ath/ath12k/wmi.c
@@ -1883,10 +1883,16 @@ int ath12k_wmi_vdev_start(struct ath12k
 				   FIELD_PREP(ATH12K_WMI_FLAG_MLO_LINK_ADD,
 					      arg->ml.link_add) |
 				   FIELD_PREP(ATH12K_WMI_FLAG_MLO_BRIDGE_LINK,
-					      arg->ml.mlo_bridge_link);
-
-		ath12k_dbg(ar->ab, ATH12K_DBG_WMI, "vdev %d start ml flags 0x%x\n",
-			   arg->vdev_id, ml_params->flags);
+					      arg->ml.mlo_bridge_link) |
+				   FIELD_PREP(ATH12K_WMI_FLAG_MLO_IEEE_LINK_IDX_VALID,
+					      1);
+
+		ml_params->ieee_link_id = arg->ml.ieee_link_id;
+
+		ath12k_dbg(ar->ab, ATH12K_DBG_WMI,
+			   "vdev %d start ml flags 0x%x ieee_link_id=%d\n",
+			   arg->vdev_id, ml_params->flags,
+			   ml_params->ieee_link_id);
 
 		ptr += sizeof(*ml_params);
 
@@ -1907,13 +1913,18 @@ int ath12k_wmi_vdev_start(struct ath12k
 			partner_info->vdev_id = arg->ml.partner_info[i].vdev_id;
 			partner_info->hw_link_id = arg->ml.partner_info[i].hw_link_id;
 			ether_addr_copy(partner_info->vdev_addr.addr, arg->ml.partner_info[i].addr);
+			partner_info->ieee_link_id = arg->ml.partner_info[i].logical_link_idx;
 
 			partner_info->flags = FIELD_PREP(ATH12K_WMI_FLAG_MLO_BRIDGE_LINK,
-							 arg->ml.partner_info[i].mlo_bridge_link);
+							 arg->ml.partner_info[i].mlo_bridge_link) |
+					      FIELD_PREP(ATH12K_WMI_FLAG_MLO_IEEE_LINK_IDX_VALID_PARTNER,
+							 1);
 
-			ath12k_dbg(ar->ab, ATH12K_DBG_WMI, "partner vdev %d hw_link_id %d macaddr %pM\n",
+			ath12k_dbg(ar->ab, ATH12K_DBG_WMI,
+				   "partner vdev %d hw_link_id %d ieee_link_id %d macaddr %pM flags 0x%x\n",
 				   partner_info->vdev_id, partner_info->hw_link_id,
-				   partner_info->vdev_addr.addr);
+				   partner_info->ieee_link_id,
+				   partner_info->vdev_addr.addr, partner_info->flags);
 			partner_info++;
 		}
 		ptr = partner_info;
--- a/drivers/net/wireless/ath/ath12k/wmi.h
+++ b/drivers/net/wireless/ath/ath12k/wmi.h
@@ -3429,10 +3429,13 @@ struct wmi_vdev_create_mlo_params {
 #define ATH12K_WMI_FLAG_MLO_LINK_ADD			BIT(8)
 #define ATH12K_WMI_FLAG_MLO_BRIDGE_PEER			BIT(10)
 #define ATH12K_WMI_FLAG_MLO_BRIDGE_LINK			BIT(14)
+#define ATH12K_WMI_FLAG_MLO_IEEE_LINK_IDX_VALID		BIT(18)
+#define ATH12K_WMI_FLAG_MLO_IEEE_LINK_IDX_VALID_PARTNER BIT(19)
 
 struct wmi_vdev_start_mlo_params {
 	u32 tlv_header;
 	u32 flags;
+	u32 ieee_link_id;
 } __packed;
 
 struct wmi_partner_link_info {
@@ -3440,6 +3443,8 @@ struct wmi_partner_link_info {
 	u32 vdev_id;
 	u32 hw_link_id;
 	struct wmi_mac_addr vdev_addr;
+	u32 flags;
+	u32 ieee_link_id;
 } __packed;
 
 struct wmi_vdev_delete_cmd {
@@ -3735,6 +3740,7 @@ struct wmi_ml_arg {
 	bool mcast_link;
 	bool link_add;
 	bool mlo_bridge_link;
+	u32 ieee_link_id;
 	u8 num_partner_links;
 	struct wmi_ml_partner_info partner_info[ATH12K_WMI_MLO_MAX_PARTNER_LINKS];
 };
